Bluetooth: Mesh: refacrtor key related functionality to use both tinycrypt and PSA based crypto libraries

alxelax opened the pull request:
This PR adds the usage of the abstract key type to be able to reuse the same BLE Mesh source code (with minor changes) for both security libraries tynicrypt and mbedTLS with PSA. For tynicrypt they are still 16 bytes array. For mbedTLS they are PSA key ID and BLE Mesh doesn't have raw key values.

alxelax commented:
1 bsim test doesn't work both crypto libraries. 
pb_adv_reprovision.sh - for tynicrypt
msg_group.sh - for mbedTLS

Balaklaka commented on subsys/bluetooth/mesh/crypto.h:
```diff
@@ -153,3 +155,12 @@ int bt_mesh_pub_key_gen(void);
+int bt_mesh_ccm_encrypt_new(const mesh_key_t key, uint8_t nonce[13],
+			const uint8_t *plaintext, size_t len, const uint8_t *aad,
+			size_t aad_len, uint8_t *enc_data, size_t mic_size);
+int bt_mesh_ccm_decrypt_new(const mesh_key_t key, uint8_t nonce[13],
```
Why do we call these functions "_new", could we just replace the current functions. This not an external API.

alxelax replied:
This is a temporary solution. I had to add them since `bt_mesh_ccm_decrypt`, `bt_mesh_ccm_encrypt` are used both by network and transport layers. I did changes only for network derivated keys. When all keys will migrate on this approach only one variant of functions will be left. All functions ...._new finally should go away,

alxelax replied:
done

Balaklaka commented on subsys/bluetooth/mesh/keys.c:
```diff
@@ -0,0 +1,37 @@
+	}
+
+	return 0;
+}
```
Are we expecting this file to grow, will it come more keys*.c for other crypto libraries? Could we just add this as a inline function in the keys.h file, so we just have the keys.h file.
What will happen when new different crypto lib comes and be need one more keys.c file.

alxelax replied:
I expect a bit more functions here. My assumption is to keep here import\export (probably something else)functionality for PSA keys. This file should be common for all crypto libraries that provide PSA API. I mean this should work for TF-M lib as well as for mbedTLS.

I'd prefer to avoid inlining this function. This is quite heavy. Inlining such heavy function in all places when keys are imported(when we migrate all keys there will be many places for import) will cause significant flash consumption.

alxelax replied:
seems like I used new types everywhere it had sense. done.

Balaklaka commented on subsys/bluetooth/mesh/crypto.h:
```diff
@@ -31,7 +33,7 @@ int bt_mesh_k1(const uint8_t *ikm, size_t ikm_len, const uint8_t salt[16],
 
 int bt_mesh_k2(const uint8_t n[16], const uint8_t *p, size_t p_len,
-	       uint8_t net_id[1], uint8_t enc_key[16], uint8_t priv_key[16]);
+	       uint8_t net_id[1], mesh_key_t *enc_key, mesh_key_t *priv_key);
```
Here the uint8_t ....[16] array is replace with mesh_key_t. We should try to use the mesh_key_t and raw_key_t def as many places as we can do.

alxelax replied:
Fully agree. As a concept, I just made changes for privacy and encryption keys since they both are derived from network key via k2.
3.8.6.3 Network key

alxelax replied:
done

Balaklaka reviewed (commented):
I think the concept is good. We should continue and fix other functions too.

PavelVPV commented on subsys/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,43 @@
+#define mesh_key_t psa_key_id_t
+#define GET_KEY_REF(key) &key
+
+int bt_mesh_import_key(enum key_type type, const raw_key_t in, mesh_key_t *out);
```
I think the implementation should belong to `crypto_tc.c` and `crypto_mtls.c`.

alxelax replied:
Moved implementation related to mbedtls into `crypto_mtls.c`. Left inline fucntions related to tinycrypt in header. There is no sense to move it in the c file since it will create jus recall `memcpy` or `memcmp`.

PavelVPV reviewed (commented):
Generally looks good. Though I would really like to get rid of tinycrypt support since it is being removed. I'll try to clarify what is expected in future.

I have one general question. Since the encryption and privacy keys are derived from a NetKey using know algorithm and inputs, shouldn't the NetKey itself be stored in PSA? Same for DevKey, AppKeys, IdentityKey and BeaconKey. This will though affect functions in `crypto.c`.

PavelVPV commented:
> Generally looks good. Though I would really like to get rid of tinycrypt support since it is being removed. I'll try to clarify what is expected in future.

Seems like tinycrypt may stay for quite long time until alternative is found :(

Balaklaka commented on tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/replay_cache/replay_attack.sh:
```diff
@@ -4,6 +4,10 @@
+# Tests must be added in pairs and in sequence.
+# First test: saves replay cache; second test: verifies it.
+
 conf=prj_pst_conf
```
Test added in pair, does that mean pairs inside the RunTest or that there has to be a pair of RunTest. If the first it true we should splitt this into two .sh files. It the second is true the test name should have something about verify.

alxelax replied:
Test description consists all necessary information about what it does and what is structure. `RunTest` is function that is called twice in this script. I'm pretty sure we do not need to describe bash syntax and how it works :)
Additionally, I do not see reason to make long cumbersome names.

Balaklaka replied:
I did not see it right away from the text and wonder if there was just placed two tests together. But it is not important.

alxelax replied:
```
# Tests must be added in pairs and in sequence.
# First test: saves replay cache; second test: verifies it.
```
Isn't this description shows the reason?

Balaklaka commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_replay_cache.c:
```diff
@@ -79,6 +80,7 @@ static void rx_ended(uint8_t *data, size_t len)
 
 static void test_tx_immediate_replay_attack(void)
 {
+	stored_keys_clear();
```
Could the 2 clear func and setup go into a init/setup function?

alxelax replied:
Actually no because some tests within test suite require them some not. There is no common initialization for them. It is exactly how we emulate restart with stored persistent data.

Balaklaka commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+#define INVALID_KEY_VAL  PSA_KEY_ID_NULL
+#define KEY_REF(key) &key
+
+int bt_mesh_import_key(enum bt_mesh_key_type type,
```
This is for all these functions. Should not the action be behind the name.
```suggestion
int bt_mesh_key_import(enum bt_mesh_key_type type,
```

Balaklaka commented on modules/mbedtls/Kconfig:
```diff
@@ -239,4 +239,34 @@ config MBEDTLS_PSA_EXTERNAL_ENTROPY
+	  Enable the emulation of the Platform Security Architecture
+	  Internal Trusted Storage (PSA ITS) over files.
+
+config MBEDTLS_FS_IO
```
Should this include _PAS in the name, and maybe _POSIX.
```suggestion
config MBEDTLS_PSA_FS_POSIX_IO
```

alxelax replied:
These are mbedTLS configuration options. We shouldn't touch naming at all since it will break mbedTLS implementation.

Balaklaka commented on modules/mbedtls/Kconfig:
```diff
@@ -239,4 +239,34 @@ config MBEDTLS_PSA_EXTERNAL_ENTROPY
+	  volatile key, or a persistent key which is loaded temporarily by the
+	  library as part of a crypto operation in flight.
+
+config MBEDTLS_PSA_CRYPTO_STORAGE_C
```
Why is _C at the end?

alxelax replied:
No idea, in general, mbedTLS configuration approach is quite cumbersome for me too.

Balaklaka commented on modules/mbedtls/Kconfig:
```diff
@@ -239,4 +239,34 @@ config MBEDTLS_PSA_EXTERNAL_ENTROPY
+	  Restrict the PSA library to supporting a maximum amount of simultaneously
+	  loaded keys. A loaded key is a key stored by the PSA Crypto core as a
+	  volatile key, or a persistent key which is loaded temporarily by the
+	  library as part of a crypto operation in flight.
```
This was a heavy sentence. And what does "in flight" mean.

alxelax replied:
Just copy-pasted option description from mbedTLS sources. Do you think it is worth to rephrase?

Balaklaka commented on modules/mbedtls/Kconfig:
```diff
@@ -239,4 +239,34 @@ config MBEDTLS_PSA_EXTERNAL_ENTROPY
+
+config MBEDTLS_PSA_CRYPTO_STORAGE_C
+	bool "The Platform Security Architecture persistent storage"
+	depends on MBEDTLS_PSA
```
Just a question: Should this be default y? So storage is enabled by default.

alxelax replied:
I guess this should be reworked (there is separate JIRA ticket for this: https://projecttools.nordicsemi.no/jira/browse/NCSDK-16568). I added mbedTLS couple of month ago when there was no any PSA mbedTLS support. While we develop PSA support in BLE Mesh it has been added in upstream on another manner but not fully (there are a lot of options those are required for us).

Balaklaka commented on subsys/bluetooth/mesh/app_keys.c:
```diff
@@ -474,9 +501,9 @@ int bt_mesh_keys_resolve(struct bt_mesh_msg_ctx *ctx,
 			}
 
-			*app_key = node->dev_key;
+			*app_key = &node->dev_key;
```
Why is not KEY_REF used here? There is some places I was wondering about this, but I only ask the question once. Or why do we need KEY_REF?

alxelax replied:
removed KEY_REF. I followed @PavelVPV  suggestion and put everything inside union. KEY_REF is not needed anymore.

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+	DEVICE_KEY
+};
+
+typedef uint8_t bt_mesh_raw_key_t[16];
```
I see no need in a separate type for the raw key. It doesn't give anything other than confusion of how it can be used in code. Pure array gives more clarity and less code refactoring.

PavelVPV commented on include/zephyr/bluetooth/mesh/cdb.h:
```diff
@@ -30,11 +31,11 @@ enum {
 
 struct bt_mesh_cdb_node {
-	uint8_t  uuid[16];
+	uint8_t uuid[16];
```
Nit-picking: Suggest to not change formatting if not needed.

alxelax replied:
I removed formatting at all due to each type renaming it should be reformatted. I guess what I have now is correct and will not cause more formatting in the future.

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+	return 0;
+}
+
+static inline int bt_mesh_check_key(const bt_mesh_raw_key_t raw_key, bt_mesh_key_t mesh_key)
```
`check` sounds more like checking key for validity. I think `compare` more close to what the function does:
```suggestion
static inline int bt_mesh_cmp_key(const bt_mesh_raw_key_t raw_key, bt_mesh_key_t mesh_key)
```

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+		       bt_mesh_key_t *out);
+int bt_mesh_export_key(bt_mesh_raw_key_t out, const bt_mesh_key_t in);
+int bt_mesh_destroy_key(bt_mesh_key_t key);
+int bt_mesh_check_key(const bt_mesh_raw_key_t raw_key, bt_mesh_key_t mesh_key);
```
Suggestion to following pattern: `bt_mesh_key_<action>` to align with other code. E.g, `bt_mesh_key_export`.

PavelVPV commented on subsys/bluetooth/mesh/app_keys.c:
```diff
@@ -106,8 +106,8 @@ static void store_app_key(uint16_t app_idx)
 
-	memcpy(key.val[0], app->keys[0].val, 16);
-	memcpy(key.val[1], app->keys[1].val, 16);
+	memcpy(KEY_REF(key.val[0]), KEY_REF(app->keys[0].val), sizeof(bt_mesh_key_t));
```
Will it make sense to wrap `memcpy` to a function, e.g, `bt_mesh_key_copy`? Though this won't be needed if `KEY_REF` is removed.

alxelax replied:
The packed data are used in some places with `memcpy` (for example in cdb). `memcpy` is designed to recognize packed attribute during compile time and build-in correct code. If separate function is implemented then unaligned access hardfault will be possible since function body will lose attribute.
Using memcpy as hardfault safe here.

PavelVPV replied:
memcpy is not needed incase of mbedtls as simple assignment can be used. For tinycrypt the function could be defined in header. In this case I guess compiler will manage to inline it correctly. But since you changing this code this comment may not be relevant.

alxelax replied:
Exactly a simple assignment will generate unaligned access hardfault. Since type is more than 1 byte the compiler will use multibyte instructions to optimize access that might cause assertion (depends how local variable will be located in the stack). Btw. the compiler understands the potential problem with it and generates warnings in all such assignments (see my latest changes for app and net key restoring from cdb). I had to add one more extra copying.

PavelVPV commented on subsys/bluetooth/mesh/app_keys.c:
```diff
@@ -497,10 +524,10 @@ int bt_mesh_keys_resolve(struct bt_mesh_msg_ctx *ctx,
 	if ((*sub)->kr_phase == BT_MESH_KR_PHASE_2 && app->updated) {
 		*aid = app->keys[1].id;
-		*app_key = app->keys[1].val;
+		*app_key = &app->keys[1].val;
```
```suggestion
		*app_key = KEY_REF(app->keys[1].val);
```

alxelax replied:
removed KEY_REF at all

PavelVPV commented on subsys/bluetooth/mesh/app_keys.c:
```diff
@@ -474,9 +501,9 @@ int bt_mesh_keys_resolve(struct bt_mesh_msg_ctx *ctx,
 			}
 
-			*app_key = node->dev_key;
+			*app_key = &node->dev_key;
```
```suggestion
			*app_key = KEY_REF(node->dev_key);
```

alxelax replied:
removed KEY_REF at all

PavelVPV commented on subsys/bluetooth/mesh/cdb.c:
```diff
@@ -858,7 +858,7 @@ void bt_mesh_cdb_node_del(struct bt_mesh_cdb_node *node, bool store)
 
 	node->addr = BT_MESH_ADDR_UNASSIGNED;
-	memset(node->dev_key, 0, sizeof(node->dev_key));
+	memset(&node->dev_key, 0, sizeof(node->dev_key));
```
```suggestion
	memset(KEY_REF(node->dev_key), 0, sizeof(node->dev_key));
```

PavelVPV replied:
Is this operation valid though? Shouldn't the key be destroyed?

PavelVPV replied:
Or shouldn't it be `node->dev_key = INVALID_KEY_VAL`?

alxelax replied:
removed removed KEY_REF as well as INVALID_KEY_VAL. INVALID_KEY_VAL is zero in PSA standard. Btw there is no guarantee it will not be changed in some revision of specification. It is Ok for now.

PavelVPV commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -223,7 +225,8 @@ void bt_mesh_reset(void)
 
-	(void)memset(bt_mesh.dev_key, 0, sizeof(bt_mesh.dev_key));
+	bt_mesh_destroy_key(bt_mesh.dev_key);
+	memset(&bt_mesh.dev_key, 0, sizeof(bt_mesh.dev_key));
```
```suggestion
	memset(KEY_REF(bt_mesh.dev_key), 0, sizeof(bt_mesh.dev_key));
```
Shouldn't it be `bt_mesh.dev_key = INVALID_KEY_VAL`?

alxelax replied:
removed INVALID_KEY_VAL

PavelVPV commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -930,11 +930,11 @@ static int net_set(const char *name, size_t len_rd, settings_read_cb read_cb,
 	}
 
-	memcpy(bt_mesh.dev_key, net.dev_key, sizeof(bt_mesh.dev_key));
+	memcpy(&bt_mesh.dev_key, &net.dev_key, sizeof(bt_mesh.dev_key));
```
```suggestion
	memcpy(KEY_REF(bt_mesh.dev_key), KEY_REF(net.dev_key), sizeof(bt_mesh.dev_key));
```

alxelax replied:
removed KEY_REF

PavelVPV commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -920,7 +919,8 @@ static int net_set(const char *name, size_t len_rd, settings_read_cb read_cb,
 		bt_mesh_comp_unprovision();
-		(void)memset(bt_mesh.dev_key, 0, sizeof(bt_mesh.dev_key));
+		bt_mesh_destroy_key(bt_mesh.dev_key);
+		memset(&bt_mesh.dev_key, 0, sizeof(bt_mesh.dev_key));
```
```suggestion
		memset(KEY_REF(bt_mesh.dev_key), 0, sizeof(bt_mesh.dev_key));
```

PavelVPV replied:
Or `= INVALID_KEY_VAL`.

alxelax replied:
removed both KEY_REF and INVALID_KEY_VAL

PavelVPV commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -399,29 +399,28 @@ static inline bool is_pb_gatt(void)
 
 	err = bt_mesh_session_key(bt_mesh_prov_link.dhkey,
-				  bt_mesh_prov_link.prov_salt, session_key);
+				  bt_mesh_prov_link.prov_salt, &session_key);
```
```suggestion
				  bt_mesh_prov_link.prov_salt, KEY_REF(session_key));
```

alxelax replied:
removed KEY_REF

PavelVPV commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -440,15 +439,13 @@ static void prov_data(const uint8_t *data)
 
 	err = bt_mesh_dev_key(bt_mesh_prov_link.dhkey,
-			      bt_mesh_prov_link.prov_salt, dev_key);
+			      bt_mesh_prov_link.prov_salt, &dev_key);
```
```suggestion
			      bt_mesh_prov_link.prov_salt, KEY_REF(dev_key));
```

alxelax replied:
removed KEY_REF

PavelVPV commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -473,7 +470,13 @@ static void prov_data(const uint8_t *data)
 	}
 
-	err = bt_mesh_provision(pdu, net_idx, flags, iv_index, addr, dev_key);
+	err = bt_mesh_import_key(NETWORK_KEY, pdu, &net_key);
```
```suggestion
	err = bt_mesh_import_key(NETWORK_KEY, pdu, KEY_REF(net_key));
```

alxelax replied:
removed KEY_REF

PavelVPV commented on subsys/bluetooth/mesh/provisioner.c:
```diff
@@ -414,20 +414,19 @@ static void send_prov_data(void)
 
 	err = bt_mesh_session_key(bt_mesh_prov_link.dhkey,
-				  bt_mesh_prov_link.prov_salt, session_key);
+				  bt_mesh_prov_link.prov_salt, &session_key);
```
```suggestion
				  bt_mesh_prov_link.prov_salt, KEY_REF(session_key));
```

alxelax replied:
removed KEY_REF

PavelVPV commented on subsys/bluetooth/mesh/provisioner.c:
```diff
@@ -414,20 +414,19 @@ static void send_prov_data(void)
 	PROV_BUF(pdu, PDU_LEN_DATA);
 	struct bt_mesh_cdb_subnet *sub;
-	uint8_t session_key[16];
+	bt_mesh_raw_key_t netkey;
```
Minor: `netkey` -> `net_key`

PavelVPV commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -10,6 +10,7 @@
 #include <zephyr/sys/util.h>
 
 #include <zephyr/settings/settings.h>
+#include <zephyr/bluetooth/mesh.h>
```
What is this header added?

alxelax replied:
mesh.h has all public API headers. Now settings.c requires key.h

PavelVPV commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1171,7 +1176,12 @@ static int cmd_net_key_add(const struct shell *shell, size_t argc, char *argv[])
 			}
 
-			memcpy(subnet->keys[0].net_key, key_val, 16);
+			if (bt_mesh_import_key(NETWORK_KEY, key_val, &subnet->keys[0].net_key)) {
```
```suggestion
			if (bt_mesh_import_key(NETWORK_KEY, key_val, KEY_REF(subnet->keys[0].net_key))) {
```

alxelax replied:
removed KEY_REF

PavelVPV commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1336,7 +1351,14 @@ static int cmd_app_key_add(const struct shell *shell, size_t argc, char *argv[])
 
-			memcpy(app_key->keys[0].app_key, key_val, 16);
+			if (bt_mesh_import_key(APPLICATION_KEY, key_val,
+					&app_key->keys[0].app_key)) {
```
```suggestion
					KEY_REF(app_key->keys[0].app_key))) {
```

alxelax replied:
removed KEY_REF

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_access.c:
```diff
@@ -303,9 +303,21 @@ static int test_msg_ne_handler(struct bt_mesh_model *model,
 static void provision(uint16_t addr)
 {
+	bt_mesh_key_t mesh_netkey;
+	bt_mesh_key_t mesh_devkey;
```
Nit-picking:
```suggestion
	bt_mesh_key_t mesh_net_key;
	bt_mesh_key_t mesh_dev_key;
```

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+	CMAC_KEY,
+	NETWORK_KEY,
+	APPLICATION_KEY,
+	DEVICE_KEY
```
`BT_MESH_` or `BT_MESH_KEY_` prefix is required.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:
```diff
@@ -364,7 +364,7 @@ static void beacon_create(struct net_buf_simple *buf, const uint8_t *net_key, ui
 	}
 
-	err = bt_mesh_beacon_key(net_key, beacon_key);
+	err = bt_mesh_beacon_key(net_key, &beacon_key);
```
```suggestion
	err = bt_mesh_beacon_key(net_key, KEY_REF(beacon_key));
```

alxelax replied:
removed KEY_REF

PavelVPV commented on subsys/bluetooth/mesh/crypto.h:
```diff
@@ -12,101 +12,75 @@ struct bt_mesh_sg {
+int bt_mesh_k4(const bt_mesh_raw_key_t n, uint8_t out[1]);
 
-int bt_mesh_id128(const uint8_t n[16], const char *s, uint8_t out[16]);
+int bt_mesh_id128(const uint8_t n[16], const char *s,
```
Either `n` should have type `bt_mesh_raw_key_t`, or as I suggested, `bt_mesh_raw_key_t` not needed as a type.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c:
```diff
@@ -376,10 +379,26 @@ static void device_setup(void)
 
-	return bt_mesh_provision(test_netkey, test_netkey_idx, test_flags, test_ividx, TEST_ADDR,
-				 test_devkey);
+	err = bt_mesh_import_key(NETWORK_KEY, test_netkey, &mesh_test_netkey);
```
```suggestion
	err = bt_mesh_import_key(NETWORK_KEY, test_netkey, KEY_REF(mesh_test_netkey));
```

alxelax replied:
removed KEY_REF

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c:
```diff
@@ -456,12 +493,24 @@ static void test_provisioning_data_load(void)
-	if (memcmp(dkey, test_devkey, sizeof(test_devkey))) {
+	if (bt_mesh_export_key(devkey, *dkey)) {
+		FAIL("Unable to export device key from crypto library");
+	} else {
```
No need for `else`, because the execution will stop in `FAIL`.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c:
```diff
@@ -456,12 +493,24 @@ static void test_provisioning_data_load(void)
 	}
 
-	if (memcmp(dkey, test_devkey, sizeof(test_devkey))) {
+	if (bt_mesh_export_key(devkey, *dkey)) {
```
If possible, it will be good to print error code as well.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/distribute_keyid.h:
```diff
@@ -0,0 +1,17 @@
+#if defined CONFIG_BT_MESH_USES_MBEDTLS_PSA
+void stored_keys_clear(void);
+#else
+static inline void stored_keys_clear(void)
```
https://github.com/zephyrproject-rtos/zephyr-bt-sig/pull/328#discussion_r989773331 :)

alxelax replied:
This is header. What I said is the truth for *.c files. If remove inlining here it'll cause syntax error.

PavelVPV replied:
It won't cause any syntax error by standard. However it may suppress `defined but not used` warning.

alxelax replied:
It will cause a situation when you get several function bodies with the same name via reincluding header. :)
Finally, it causes a compiler or linker error (do not remember which one will generate it).
For header functions `static inline` is mandatory or at least wishible to avoid this. For source file functions it doesn't have sense.

PavelVPV replied:
> It will cause a situation when you get several function bodies with the same name via reincluding header. :)

No, it won't as long as header is guarded.

> Finally, it causes a compiler or linker error (do not remember which one will generate it).

No, this is not how static keyword works.

alxelax replied:
> > It will cause a situation when you get several function bodies with the same name via reincluding header. :)
> 
> No, it won't as long as header is guarded.
> 

Some headers are guarded some aren't. Common and good practice to not put function bodies in the headers other than static inline functions. It protects from what I described above. I do not see any reason to do oppositely to worldwide practice.  

> > Finally, it causes a compiler or linker error (do not remember which one will generate it).
> 
> No, this is not how static keyword works.

Hm. My sentences are correct in the pair. Both describe exact compiler\linker behavior. In general, it doesn't matter function is static or isn't. Without guarding the issue will be. Static keyword will not help. No sense to discuss the second part without the first.

PavelVPV replied:
Your header is guarded. So no issue for this case. But generally I agree if you define 2 functions with the same name in one unit, this will cause linker error. But it is not the case for this.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/distribute_keyid.c:
```diff
@@ -0,0 +1,69 @@
+		err = 0;
+	}
+
+	return err;
```
Suggest to move successful case out of `if`:
```suggestion
	unsigned int unique = get_device_nbr();

	unique <<= 8;

	if (!IN_RANGE(key_id - unique, PSA_KEY_ID_USER_MIN,
			PSA_KEY_ID_USER_MIN + KEY_ID_RANGE_SIZE - 1)) {
		return -EIO;
	}
	
	LOG_INF("key id %d is freed", pst_key_id[key_id - PSA_KEY_ID_USER_MIN]);
	pst_key_id[key_id - PSA_KEY_ID_USER_MIN] = PSA_KEY_ID_NULL;

	return 0;
```

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/distribute_keyid.c:
```diff
@@ -0,0 +1,69 @@
+{
+	unsigned int unique = get_device_nbr();
+
+	unique <<= 8;
```
Could it be wrapped into a macro?
```
#define DEV_ID (get_device_nbr() << 8)
```

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/distribute_keyid.c:
```diff
@@ -0,0 +1,69 @@
+
+	unique <<= 8;
+
+	bt_mesh_crypto_init();
```
I'd rather move the initialization out of this function as this is not what you expect from a function call `_clear`.

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_MESH_KEYS_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_MESH_KEYS_H_
+
+#include <inttypes.h>
```
What is the purpose if `inttypes.h`? Why not `stdint.h`?

PavelVPV reviewed (changes requested):
I did a high-level review of the code and have identified some issues:

1. Is `KEY_REF` really needed? For both cases, PSA and tinycrypt, using `&` is valid. More over, I see in many places `KEY_REF` is missed, but the code still work (because it is valid). I think it will be easy to forget to use `KEY_REF`, when writing or adding new code. So I'd rather write the code where this macro is not needed, and, hopefully, in this particular case, nothing needs to be changed.
2. There are few cases where `INVALID_KEY_VAL` is used. I see `INVALID_KEY_VAL` creates issues when a key needs to be reset: can the key be memset or can it be assigned to `INVALID_KEY_VAL`? It would be better to replace it with NULL, e.g. using pointer to `bt_mesh_key_t` as a function argument in `bt_mesh_app_key_set` and `bt_mesh_subnet_set`.
3. If we support tinycrypt in bsim tests, we need to have a separate target that will run with it. Otherwise, if it is not run, its support is of no use. However, need to consider which tests we need to run with it so that it will not significantly affect CI time. I think we can to create a separate task to resolve this.

I'll do a detailed review now.

alxelax commented:
> I did a high-level review of the code and have identified some issues:
> 
> 1. Is `KEY_REF` really needed? For both cases, PSA and tinycrypt, using `&` is valid. More over, I see in many places `KEY_REF` is missed, but the code still work (because it is valid). I think it will be easy to forget to use `KEY_REF`, when writing or adding new code. So I'd rather write the code where this macro is not needed, and, hopefully, in this particular case, nothing needs to be changed.

My commit comment when I added `KEY_REF` usage. I also thought it is not necessary but actually there is some problem with the array type.
```
Despite pointer on the array type and array itself are the same
address the compiler handles them on the different manner.
Copying after `addres of` operator doesn't cause copying array
but some dereferencing memory. That caused the situation when
bsim tests worked but used garbage instead of real keys for
tinycrypt library. Using macro for getting correct reference
to array by name instead by pointer to array solves the issue.
Behavior for mbedtls library wasn't changed.
```

If I missed somewhere usage `KEY_REF` this is a bug. Please point to places, I'll fix them.

PavelVPV commented:
> My commit comment when I added `KEY_REF` usage. I also thought it is not necessary but actually there is some problem with the array type.
> 
> ```
> Despite pointer on the array type and array itself are the same
> address the compiler handles them on the different manner.
> Copying after `addres of` operator doesn't cause copying array
> but some dereferencing memory. That caused the situation when
> bsim tests worked but used garbage instead of real keys for
> tinycrypt library. Using macro for getting correct reference
> to array by name instead by pointer to array solves the issue.
> Behavior for mbedtls library wasn't changed.
> ```

Can this be solved by wrapping up array into a structure? In function arguments use pointer to the key.

> If I missed somewhere usage `KEY_REF` this is a bug. Please point to places, I'll fix them.

Posted as comments.

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+		       bt_mesh_key_t *out);
+int bt_mesh_export_key(bt_mesh_raw_key_t out, const bt_mesh_key_t in);
+int bt_mesh_destroy_key(bt_mesh_key_t key);
+int bt_mesh_check_key(const bt_mesh_raw_key_t raw_key, bt_mesh_key_t mesh_key);
```
Is there a need in these 3 functions to be public? I can imaging `bt_mesh_import_key` needs to be public for self-provisioning, but what is the use case for the other 3?

alxelax replied:
What is your suggestion?

PavelVPV replied:
Can we hide them?

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -357,3 +296,145 @@ __weak psa_status_t mbedtls_psa_external_get_random(mbedtls_psa_external_random_
+
+int bt_mesh_check_key(const bt_mesh_raw_key_t raw_key, bt_mesh_key_t mesh_key)
+{
+	uint8_t out[16];
```
Why is not this defined to bt_mesh_raw_key_t?

alxelax replied:
Removed `bt_mesh_raw_key_t` either

Balaklaka commented on subsys/bluetooth/mesh/subnet.c:
```diff
@@ -309,29 +334,29 @@ static int net_keys_create(struct bt_mesh_subnet_keys *keys,
 #endif /* GATT_PROXY */
 
-	err = bt_mesh_beacon_key(key, keys->beacon);
+	err = bt_mesh_beacon_key(key, &keys->beacon);
```
&->KEY_REF

alxelax replied:
removed

PavelVPV commented:
> > If I missed somewhere usage `KEY_REF` this is a bug. Please point to places, I'll fix them.
> 
> Posted as comments.

I see now that some of my comments regarding `KEY_REF` maybe incorrect. This makes me feel we need to try something else to solve this correctly.

PavelVPV commented:
I don't see where lpn and friend credentials are destroyed upon `BT_MESH_KEY_DELETED`, `BT_MESH_KEY_UPDATED` and when friendship is terminated. Or do I miss something?

Balaklaka commented on subsys/bluetooth/mesh/app_keys.h:
```diff
@@ -41,7 +41,7 @@ int bt_mesh_app_key_set(uint16_t app_idx, uint16_t net_idx,
 int bt_mesh_keys_resolve(struct bt_mesh_msg_ctx *ctx,
 			 struct bt_mesh_subnet **sub,
-			 const uint8_t **app_key, uint8_t *aid);
+			 const bt_mesh_key_t **app_key, uint8_t *aid);
```
Is it correct to have a **app_key here, when we changed the type?

alxelax replied:
yep, it is correct

Balaklaka commented on subsys/bluetooth/mesh/crypto.c:
```diff
@@ -68,7 +76,7 @@ int bt_mesh_k2(const uint8_t n[16], const uint8_t *p, size_t p_len,
 	}
 
-	err = bt_mesh_aes_cmac_one(salt, n, 16, t);
+	err = bt_mesh_aes_cmac_one_raw_key(salt, n, 16, t);
```
Why is not salt/t/out defined as bt_mesh_raw_key_t? As that is what bt_mesh_aes_cmac_one_raw_key takes as input. There are more of these in this file.

alxelax replied:
salt is not key. This is separate entity according to encryption algorithms

Balaklaka commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/mesh_test.c:
```diff
@@ -182,16 +182,32 @@ const struct bt_mesh_comp comp = {
 
-const uint8_t test_net_key[16] = { 1, 2, 3 };
+const bt_mesh_raw_key_t test_net_key = { 1, 2, 3 };
 const uint8_t test_app_key[16] = { 4, 5, 6 };
```
```suggestion
const bt_mesh_raw_key_t test_app_key = { 4, 5, 6 };
```

alxelax replied:
removed `bt_mesh_raw_key_t`

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,83 @@
+
+#elif defined CONFIG_BT_MESH_USES_TINYCRYPT
+
+#define bt_mesh_key_t bt_mesh_raw_key_t
```
Just wonder, what is the benefit of defining a macro instead of using typedef?

alxelax replied:
Compliance checker warning:
```
 WARNING:NEW_TYPEDEFS: do not add new typedefs
```

PavelVPV replied:
Hmm, maybe we can bypass this warning when create final PR.

alxelax replied:
aligned source code to compliance requirements. Looks cumbersome but no warnings now :)

alxelax commented:
> > > If I missed somewhere usage `KEY_REF` this is a bug. Please point to places, I'll fix them.
> > 
> > 
> > Posted as comments.
> 
> I see now that some of my comments regarding `KEY_REF` maybe incorrect. This makes me feel we need to try something else to solve this correctly.

I followed your suggestion and put key into union. I slightly reworked usage. I'll provide it on one more round of review when it is fully ready.

alxelax commented:
> I don't see where lpn and friend credentials are destroyed upon `BT_MESH_KEY_DELETED`, `BT_MESH_KEY_UPDATED` and when friendship is terminated. Or do I miss something?

Good catch. Really overlooked these changes.

alxelax committed:
Bluetooth: Mesh: refactor mesh to use both tinycrypt and psa based crypto

Mesh key type has been added to be able to choose the different
key representation for different security libraries.
The type as well as some functionality related to Mesh key
management have been added as a public API.
If tynicrypt is chosen then keys have representation
as 16 bytes array. If mbedTLS with PSA is used then keys are
the PSA key id. Raw value is not kept within BLE Mesh stack
for mbedTLS. Keys are imported into security library
and key ids are gotten back. This refactoring have been done
for network(including all derivated keys), application,
device and session keys.
Also PR adds ability to build BLE Mesh bsim tests
with both tynicrypt and mbedTLS crypto libraries.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

PavelVPV commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -357,3 +296,145 @@ __weak psa_status_t mbedtls_psa_external_get_random(mbedtls_psa_external_random_
+		break;
+	case NETWORK_KEY:
+		if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+			psa_key_id_t key_id = bt_mesh_user_keyid_alloc();
```
When a device is provisioned or a new network key is added, the key is imported by `net_keys_create` function and `bt_mesh_user_keyid_alloc` is called:
```
bt_mesh_subnet_add -> net_keys_create (import = true)
                            |-> msg_cred_create -> bt_mesh_k2 -> bt_mesh_import_key (Encryption and Privacy keys)
                            |-> bt_mesh_import_key (net key) -> *bt_mesh_user_keyid_alloc*
                            |-> bt_mesh_import_key (identity key)
                            \-> bt_mesh_import_key (beacon key)
```

But when the key id is restored after reboot, the network key is not imported and `bt_mesh_user_keyid_alloc` is not called:
```
net_key_set -> bt_mesh_subnet_set -> subnet_key_set -> net_keys_create (import = false)
```
Unless I miss something, this breaks consistency of `pst_key_id` array between reboots and incorrect consequent id allocation.

Could we avoid having `pst_key_id` array and access `app_keys.c` and `subnet.c` to check if id is occupied? And shouldn't we declare publicly the range of ids? In case some other model, for example, GATT?, may also use PSA for storing keys.

alxelax replied:
Good point. I don't import in `bt_mesh_subnet_set ` by design because my assumption was it has been allocated by the application in advance. Case when it is restored from cdb I missed. This is the obvious issue. If you don't have an objection I'd suggest to solve this as a separate task. It seems quite an autonomic task that could be done separately.

PavelVPV replied:
To be addressed in NCSDK-17765

PavelVPV reviewed (commented):
One more finding:

In few places we export keys from the crypto storage to do some things, like generating ids at the boot or doing key comparison when adding a new key. This design is driven by the previous implementation where we keys were stored explicitly by mesh and this was fine. But with PSA I think the more correct would be to not export keys. From what I see the key exporting can be avoided for nodes. These are the places where the key is exported and how exporting can be avoided:
- When generating AppKeyId in `app_id_set`.
    - Possible solution: store id persistently when application key is added.
- In `subnet_key_set`.
    - Possible solution: use the key derivation API with passing psa_key_id_t of the network key.
- In `bt_mesh_friend_cred_create`.
    - Possible solution: use the key derivation API with passing psa_key_id_t of the network key.
- When adding new application or network key.
    - Possible solution: use key hash for comparison or compare keys by encrypting the same text.

This is though not avoidable for provisioner since it needs to send keys explicitly.

PavelVPV commented:
> One more finding:
> 
> In few places we export keys from the crypto storage to do some things, like generating ids at the boot or doing key comparison when adding a new key. This design is driven by the previous implementation where we keys were stored explicitly by mesh and this was fine. But with PSA I think the more correct would be to not export keys. From what I see the key exporting can be avoided for nodes. These are the places where the key is exported and how exporting can be avoided:
> 
> * When generating AppKeyId in `app_id_set`.
>   
>   * Possible solution: store id persistently when application key is added.
> * In `subnet_key_set`.
>   
>   * Possible solution: use the key derivation API with passing psa_key_id_t of the network key.
> * In `bt_mesh_friend_cred_create`.
>   
>   * Possible solution: use the key derivation API with passing psa_key_id_t of the network key.
> * When adding new application or network key.
>   
>   * Possible solution: use key hash for comparison or compare keys by encrypting the same text.
> 
> This is though not avoidable for provisioner since it needs to send keys explicitly.

I think this should not be done in this PR and be addressed later.

alxelax commented:
@PavelVPV, @Balaklaka I see a couple of comments should be converted to tasks. I'll gather them and provide links here.

alxelax commented:
> > I don't see where lpn and friend credentials are destroyed upon `BT_MESH_KEY_DELETED`, `BT_MESH_KEY_UPDATED` and when friendship is terminated. Or do I miss something?
> 
> Good catch. Really overlooked these changes.

fixed

alxelax commented:
> @PavelVPV, @Balaklaka I see a couple of comments should be converted to tasks. I'll gather them and provide links here.
https://projecttools.nordicsemi.no/jira/browse/NCSDK-16568
https://projecttools.nordicsemi.no/jira/browse/NCSDK-17765
https://projecttools.nordicsemi.no/jira/browse/NCSDK-17766

PavelVPV commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,62 @@
+#endif
+
+enum bt_mesh_key_type {
+	BT_MESH_ECB_KEY,
```
I would really like to move `_KEY` to prefix:
`BT_MESH_KEY_<type>`, or even `BT_MESH_KEY_TYPE_<type>` (to follow enum name). This makes easier search for key usages in the code.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c:
```diff
@@ -42,7 +44,9 @@ static int test_ividx = 0x123456;
 static uint8_t test_flags;
 static uint8_t test_netkey_idx = 0x77;
 static uint8_t test_netkey[16] = { 0xaa };
+static struct bt_mesh_key mesh_test_netkey;
```
Nit-picking (just to align with other code):
```suggestion
static struct bt_mesh_key mesh_test_net_key;
```
There are few more cases where `netkey` or `devkey` are used without underscore.

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -221,8 +146,23 @@ int bt_mesh_aes_cmac(const uint8_t key[16], struct bt_mesh_sg *sg,
 	}
 
 end:
```
Why not remove the end: and just return where it is goto end:

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -357,3 +297,145 @@ __weak psa_status_t mbedtls_psa_external_get_random(mbedtls_psa_external_random_
+
+__weak int bt_mesh_user_keyid_free(psa_key_id_t key_id)
+{
+	int err = -EIO;
```
Why not remove this end and set return 0 and return -EIO directly.

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -221,8 +146,23 @@ int bt_mesh_aes_cmac(const uint8_t key[16], struct bt_mesh_sg *sg,
+			size_t sg_len, uint8_t mac[16])
+{
+	struct bt_mesh_key key_id;
+	int err = 0;
```
Not necessary with =0

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -357,3 +297,145 @@ __weak psa_status_t mbedtls_psa_external_get_random(mbedtls_psa_external_random_
+	}
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		err = bt_mesh_user_keyid_free(key->key);
```
```suggestion
		return bt_mesh_user_keyid_free(key->key);
```
And remove int err = 0; and return 0 at the end

Balaklaka commented on subsys/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,31 @@
+
+#elif defined CONFIG_BT_MESH_USES_TINYCRYPT
+
+static inline int bt_mesh_key_export(uint8_t out[16], const struct bt_mesh_key *in)
```
If we put this functions in crypto_tc.c we do not need the #if defined here. The *.h file then becomes very clean.

alxelax replied:
answered here: https://github.com/alxelax/zephyr/pull/6#discussion_r1011940084

Balaklaka commented on include/zephyr/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,62 @@
+	uint8_t key[16];
+};
+
+static inline int bt_mesh_key_import(enum bt_mesh_key_type type,
```
Why not put this function in crypto_tc.c and then this function can be outside the #if defined. Then the only thing inside the #if defined is the struct bt_mesh_key.

alxelax replied:
because compiler will create prebuilt instances these functions and link them as objects. That significantly decrease advantage of usage `memcpy` function. By default compiler tries roll out each usage of these function to generate the most optimal code with usage the bus widest available instruction (e.g. 4 bytes or 2 bytes for cortex). At the same it follows alignment of memory addresses for the begin of region and for the tail of region those are not possible to handle with widest bus instructions to avoid memory bus exceptions. Putting just `memcpy` inside the object will tie the compiler optimization hands.

yes, moving into source file will make header more clean but it will create less optimal code for tinycrypt comparing to initial.

Balaklaka commented on subsys/bluetooth/mesh/keys.h:
```diff
@@ -0,0 +1,31 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Missing h file #include/#define (guard). Should this file be called keys.h as we already have a keys.h. Could be called keys_internal.h instead or go to crypto.h as the functions are in crypto_mtls.c.

alxelax replied:
We do not have `internal` suffix for internal headers. They are just located in the same place as the source files. Then just including then as local headers highlights it is internal or doesn't. Even more, stack sources and `internal` headers do not include `external` headers explicitly only via `mesh.h`. There is no direct path to these headers.

No one `internal` header has guard definitions. It helps to avoid to including headers within the headers to have a deal with strange interfile dependencies. Instead of this all headers are included only in source files in a sequence they depend on each other. It allows to not bother about guards. Of course, there are a couple of headers included inside headers and these are actual bugs that sneaked through review. :(

Balaklaka commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -37,6 +37,7 @@
 #include "settings.h"
 #include "prov.h"
 #include "cfg.h"
+#include "keys.h"
```
Which keys.h file is included?

alxelax replied:
answered this question here: https://github.com/alxelax/zephyr/pull/6#discussion_r1011899656

Balaklaka commented on subsys/bluetooth/mesh/provisioner.c:
```diff
@@ -33,6 +33,7 @@
 #include "proxy.h"
 #include "prov.h"
 #include "settings.h"
+#include "keys.h"
```
Could be hard understand which keys.h file this is including.

alxelax replied:
Hm. Isn't this according to C language syntax? 
If compiler sees `#include "header.h"` then compiler looks for header file in the same folder where it is located at the current moment.
If compiler sees `#include <header.h>` then compiler looks for header file in the folders those have been provided as system or parameterized paths (via `-I`) .

alxelax replied:
This is absolutely the same as it has been done for other Mesh headers. Please take a look at `access.h`, `cdb.h` etc etc
